### DISCUSSION TITLE
No convergence for aster computation

### DISCUSSION THREAD

**Discussion by SebastienMeunier** [thelfer/tfel#379] [Q&A]
I use the StandardElastoViscoPlasticity Brick inside code_aster and sometimes, I encounter the error in STAT_NON_LINE resolution, due to the MFront file : <EXCEPTION> <DVP_2>  Erreur numérique (floating point exception).

I have discovered the existence of the keyword @UMATUseTimeSubStepping true

By any chance, might I use this keyword @UMATUseTimeSubStepping, within a code_aster command file, for time substepping please ?

Best regards

**Comment by thelfer** [comment]
@SebastienMeunier This feature is only available in the `Cast3M` interface, since local substepping degrades the quality of the consistent tangent operator. Have you tried to generate `MTest` files to analyse the trouble ?

**Reply by SebastienMeunier** [reply]
Thank you @thelfer for the quick answer. No I have not tried. I have a look at your suggestion: I try to generate MTest file

**Reply by thelfer** [reply]
I am interested. Do not hesitate to share your MFront file and your MTest file if you can !

**Comment by SebastienMeunier** [comment]
Could you just remind me how to do that please ? I just have to add an instruction @GenerateMTestFileOnFailure or something like that in my mfront file please ?

**Reply by thelfer** [reply]
Yes

```cxx
@GenerateMTestFileOnFailure  true;
```

I personnaly do this on the command line as follows:

```bash
$ mfront --obuild --interface=aster --@GenerateMTestFileOnFailure=true file.mfront
```

Then you must wrap the call to `STAT_NON_LINE` in a `try/except` block and copy all the generated `MTest` files to a local directory:

```python
try:
  STAT_NON_LINE(....)
except:
   os.system("cp *.mtest <local_directory>")
```

Note that the computation may succeed if substepping is enabled in  `STAT_NON_LIN`, you may also try to copy the `MTest` files even on success.

**Reply by SebastienMeunier** [reply]
I tried but unfortunately, with the attached mfront file, within the salome-meca container, with the following command:
`mfront-3.4.3 --obuild --interface=aster --@GenerateMTestFileOnFailure=True visc_isot.mfront `
I obtain the message:

```
Error while treating file '/home/F82953/aster/vimfr/visc_isot.mfront'
BehaviourDSLCommon::analyse: error while treating keyword '@GenerateMTestFileOnFailure' at line '1' of file 'visc_isot.mfront'.
DSLBase::treatUnknownKeyword: unknown keyword (read '@GenerateMTestFileOnFailure')
```
[visc_isot.zip](https://github.com/thelfer/tfel/files/11097370/visc_isot.zip)

I do not know what to do. Could you help me please ?

**Reply by thelfer** [reply]
Argh. Old version of `mfront`.  The following command shall work:

```bash
$ mfront --obuild --interface=aster --@AsterGenerateMTestFileOnFailure=true visc_isot.mfront
```

**Reply by SebastienMeunier** [reply]
I tried to generate mtest files but i did not succeed. I see that mtest files are generated in the temporary aster execution directory but I do not succeed to get them on my computer, as if
`os.system("cp *.mtest <local_directory>")`
is not executed :(

**Reply by thelfer** [reply]
did you try to add some debug messages, i.e. something like `print("HERE")` to see if we go inside the `except` branch
